### ISSUE TITLE
Create a Huggingface demo page

### ISSUE THREAD

**Issue by jpc** [collabora/WhisperSpeech#39]
The output quality is good enough that it would be useful to allow more folks to test our model out on Huggingface.

**Comment by itsharshitrwt** [comment]
Hi , @jpc  i m very much new to open source and want to start contributing  , I came across this repo and saw good first issue label . It will be really helpful  if you could just tell me how to work on this issue ,  where to start.
I m excited to work on this issue.

**Comment by jpc** [comment]
Hey, @Josephrp started working on this one just yesterday. Maybe you can reach out to him on the LAION Discord (link in the README, his username: Tonic) and ask if he needs any help.

**Comment by gongouveia** [comment]
hey, what is the status of this feature? I can make it work this week.
Can I help @Josephrp ?

**Comment by Josephrp** [comment]
hey tehre @gongouveia @jpc @itsharshitrwt , i'm so glad i finally caught these, with all my appologies, i'll basically be finalizing the demo today, but help is totally welcome and encouraged ! how fun.

outstanding tasks are simple :

- review and improve the demo branding (and model card ?) for whisperspeech , there's a lot of important information we want folks to have and also have that in the demo code :-)
- we'll simply the interface for sure one last and thought through refactor to make super easy to do a multilingual voice print and a multilingual text to voice
- we'll create a simple text parser and audio parser to basically allow the user to input a multilingual string , as so : <pl> bobber kurkuma <fr> oh mon dieu qu'est-ce-que c'est ? <hin>ये महत्वाकांक्षी लक्ष्य हैं and get multilingual with emotion based on whether or not an audio file is provided for the voice. 
- for V2 i think we can improve the text parser with autodetect, cant we ? 

to summarize : 

simple more powerful and better looking single screen app with a little backend flair to show off the whisperspeech library in a good way = today's goal :-)

**Comment by Josephrp** [comment]
@jpc @gongouveia @itsharshitrwt , oops forgot to share the demo link:

https://huggingface.co/spaces/Tonic/laion-whisper

it's been up already some... and i've just been kinda sharing it around lol , as i've been gearing up, but not here, yet, so sorry about that. please take a look there, you can fork that one and make PRs directly, the easiest is to coordinate on issues and make PRs accordingly, even if it takes a tiny bit more time :-)

**Comment by Josephrp** [comment]
@itsharshitrwt @gongouveia , i've updated the space name here : https://huggingface.co/spaces/Tonic/whisperspeech/

**Comment by Josephrp** [comment]
join me here : https://discord.gg/4r9akpvF?event=1200011410638393404

**Comment by gongouveia** [comment]
> * we'll simplify the interface for sure one last and thought-through refactor to make super easy to do a multilingual voice print and a multilingual text to voice

Will join as soon as possible, I believe the space is good enough already to showcase. 
However in "we'll simplify the interface for sure one last and thought-through refactor to make super easy to do a multilingual voice print and a multilingual text to voice", my personal preference is inputs in left and outputs in right side. Generate speech button is too big now, I will propose a new interface configuration :)

**Comment by Josephrp** [comment]
> > * we'll simplify the interface for sure one last and thought-through refactor to make super easy to do a multilingual voice print and a multilingual text to voice
> 
>  preference is inputs in left and outputs in right side. Generate speech button is too big now, I will propose a new interface configuration :)

that's great, dont launch forth just yet because i'm pushing the next version

**Comment by Josephrp** [comment]
i need help with that interface stuffs :-) @gongouveia

**Comment by Josephrp** [comment]
debugging this error currently (new interface is up and running :-) ) 

```bash

  File "/home/user/.local/lib/python3.10/site-packages/spaces/zero/wrappers.py", line 164, in thread_wrapper
    res = future.result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/user/app/app.py", line 62, in whisper_speech_demo
    audio_np = generate_segment_audio(text, lang, speaker_url, pipe)
  File "/home/user/.local/lib/python3.10/site-packages/spaces/zero/wrappers.py", line 91, in gradio_handler
    return task(*args, **kwargs)
  File "/home/user/app/app.py", line 39, in generate_segment_audio
    audio_data = pipe.generate(stoks, speaker_url, lang)
  File "/home/user/app/whisperspeech/pipeline.py", line 87, in generate
    return self.vocoder.decode(self.generate_atoks(text, speaker, lang=lang, cps=cps, step_callback=step_callback))
  File "/home/user/app/whisperspeech/pipeline.py", line 81, in generate_atoks
    text = text.replace("\n", " ")
AttributeError: 'Tensor' object has no attribute 'replace'
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/gradio/queueing.py", line 495, in call_prediction
    output = await route_utils.call_process_api(
  File "/home/user/.local/lib/python3.10/site-packages/gradio/route_utils.py", line 232, in call_process_api
    output = await app.get_blocks().process_api(
  File "/home/user/.local/lib/python3.10/site-packages/gradio/blocks.py", line 1561, in process_api
    result = await self.call_function(
  File "/home/user/.local/lib/python3.10/site-packages/gradio/blocks.py", line 1179, in call_function
    prediction = await anyio.to_thread.run_sync(
  File "/home/user/.local/lib/python3.10/site-packages/anyio/to_thread.py", line 56, in run_sync
    return await get_async_backend().run_sync_in_worker_thread(
  File "/home/user/.local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 2134, in run_sync_in_worker_thread
    return await future
  File "/home/user/.local/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 851, in run
    result = context.run(func, *args)
  File "/home/user/.local/lib/python3.10/site-packages/gradio/utils.py", line 678, in wrapper
    response = f(*args, **kwargs)
  File "/home/user/.local/lib/python3.10/site-packages/spaces/zero/wrappers.py", line 118, in gradio_handler
    raise res.value
AttributeError: 'Tensor' object has no attribute 'replace'

```

**Comment by Josephrp** [comment]
hey there , i refactored the functional demo code for a better example code, so i'll just make a fork for that and push to readme or something (?) @jpc with a big thanks to jose that helped out too :-)

**Comment by Josephrp** [comment]
so i made a fork here : https://github.com/Josephrp/WhisperSpeech/tree/main with the gradio demo added

**Comment by Josephrp** [comment]
@gongouveia @itsharshitrwt @jpc , open issues are in the discussions tab here : https://huggingface.co/spaces/Tonic/whisperspeech/discussions

- reformat the language list (and therefore the interface layout - one last time :-) ) 
- adding cool examples
- reach goal : slightly better parser for handling punctuation and common user input formats ;-)

**Comment by Josephrp** [comment]
@jpc just starting the final changes to the layout namely :
- display the language tags inside <tags> 
- small reformat to clean up the "tabs" , put the language tags on the bottom display=false by default + it should be renamed Language Tags or help + small improvements for the uer

afterwards i will port these changes to the demo app code here : https://github.com/collabora/WhisperSpeech/issues/39#issuecomment-1910207333

**Comment by Josephrp** [comment]
so, i do prefer the new layout and language list, just pushing to my fork and standing by for your instructions for how to proceed @jpc 

![image](https://github.com/collabora/WhisperSpeech/assets/18212928/236b9bad-c1bc-4240-ba7f-1d1800e28c6f)

**Comment by Josephrp** [comment]
@itsharshitrwt @gongouveia : see the PR here : https://github.com/collabora/WhisperSpeech/pull/63 it would be useful to work together on the outstanding issues

**Comment by gongouveia** [comment]
I will play around with it a bit today

**Comment by gongouveia** [comment]
> I will play around with it a bit today

@Josephrp I'm looking on it now.
Will work on:

1.  new examples
2.  Caching for examples.
3.  adding Queue (never tried it with gpu zero)
4.  Rewriting text, has some errors
5.  Redo Layout, it should be inputs, outputs. outputs above doesn´t make sense.
6.

**Comment by Josephrp** [comment]
> * Redo Layout, it should be inputs, outputs. outputs above doesn´t make sense.

sorry... outputs above is standard, here are three examples with outputs above :

- github issues 
- ableton
- every ai chat app ever.

just check any of these ;-) 

hope you have fun, let's connect whenever you want you can find me :-)

**Comment by gongouveia** [comment]
> > * Redo Layout, it should be inputs, outputs. outputs above doesn´t make sense.
> 
> sorry... outputs above is standard, here are three examples with outputs above :
> 
> * github issues
> * ableton
> * every ai chat app ever.
> 
> just check any of these ;-)
> 
> hope you have fun, let's connect whenever you want you can find me :-)

Got it, 
I fancy as this, whats your opinion? 
![gota](https://github.com/collabora/WhisperSpeech/assets/68733294/5dece261-586f-4c84-bcaa-d158a9ed497b)


Btw in the current version, the mic/upload audio button is messed up.
Will add examples with pairs of text/ voice tomorrow

**Comment by Josephrp** [comment]
i also noticed that, about the mic and upload icons on my 32bit windows computer replacement but i thought it was just my OS :-) 

i simply cannot imagine what i need to fix about it because it's... a gradio element i guess, nothing custom... it's a headscratcher , so i'll think with you about it . 

new layout is much cleaner, the smaller drop downs are easier to see with the text makes more sense. if we're quickly experimenting , could even be to switch it with audio upload, i mean , jsut to see if it's better/same/worse .

one suggestion, maybe, for the audio+text examples is to put some examples also on the audio upload side , more clicking on easy things is always great. (just my random thoughts! sorry if it's wrong :-) ) 

final thing : 

check the demo actually, i think it could be you, or maybe not ?  but @mrfakename basically made a really cool PR , eg : 

```python
@spaces.GPU(enable_queue=True)
def whisper_speech_demo(multilingual_text, speaker_audio):
```

this actually fixed some performance issues i was seeing with a lot of languages together. i'm very happy, so check it out in case its not your push :-)

**Comment by gongouveia** [comment]
It is not my push, 
I also checked if this gradio element, was just working badly in microsoft edge, but in chrome has same behavior.
I recently also found that the audio element behaves differently in different search engines, in chrome it does some weird audio normalization.

**Comment by Josephrp** [comment]
> It is not my push, I also checked if this gradio element, was just working badly in microsoft edge, but in chrome has same behavior. I recently also found that the audio element behaves differently in different search engines, in chrome it does some weird audio normalization.
> 
> Tell me, audio samples must be how many seconds long, 30 seconds as in whisper?

i didnt realize there was a top limit, but normally more is better, 20 seconds or more is better.

btw i'm having performance issues now on huggingface, but i think it's just huggingface, since uhm the demo was working not too long ago... might revert the changes, but i'll check it tomorrow :-)

**Comment by Josephrp** [comment]
https://huggingface.co/spaces/Tonic/whisperspeech/discussions/4 

check this discussion for the logs from an error that ... is strange + i'm investigating in the context of the recent collabora updates , help also there , is more than welcome.

**Comment by gongouveia** [comment]
> > It is not my push, I also checked if this gradio element, was just working badly in microsoft edge, but in chrome has same behavior. I recently also found that the audio element behaves differently in different search engines, in chrome it does some weird audio normalization.
> > Tell me, audio samples must be how many seconds long, 30 seconds as in whisper?
> 
> i didnt realize there was a top limit, but normally more is better, 20 seconds or more is better.


The way whisper architecture trasncribes long from audios, is chunking them in 30 seconds audio sample samples with little of overlap between chunks. 

I am lacking creativity to find audio voice samples, plus I don´t know what can be copyrighted or not tho. maybe will use some librespeech dataset samples

**Comment by Josephrp** [comment]
> > > It is not my push, I also checked if this gradio element, was just working badly in microsoft edge, but in chrome has same behavior. I recently also found that the audio element behaves differently in different search engines, in chrome it does some weird audio normalization.
> > > Tell me, audio samples must be how many seconds long, 30 seconds as in whisper?
> > 
> > 
> > i didnt realize there was a top limit, but normally more is better, 20 seconds or more is better.
> 
> The way whisper architecture trasncribes long from audios, is chunking them in 30 seconds audio sample samples with little of overlap between chunks.
> 
> I am lacking creativity to find audio voice samples, plus I don´t know what can be copyrighted or not tho. maybe will use some librespeech dataset samples

interesting ! so maybe a little session of peercoding is in order to wrap up the outstanding issues, normally chunking in 30 secords should be easy and possible + it seems like it's quite hard to record a whole 30 seconds :-) honestly i was just going to upload my own voice saying goofy and feel good stuff :-) i think it's too late to protect my voice print already...

**Comment by Josephrp** [comment]
hmmm @gongouveia , tonight maybe it would be nice to also re organise the interface : currently there's audio upload and text in a row below the output... but... you're right, it doesnt make sense... maybe also it could look a bit more slick imho, so by reorganising the upload to it's own row perhaps, and perhaps "on top of" text input , thinking that it could be combined with the language tags for extra information in a good way, perhaps as rows inside a single accordion .

**Comment by gongouveia** [comment]
You mean peer coding in discord?
If thats the case I believe our time zone is very different

**Comment by Josephrp** [comment]
seems that:

- the parser is working,
- the audio quality has been improved
- interface has be refactored/arranged

...

therefore it's now interesting / important to move towards finalizing this contribution , perhaps as a notebook to launch a gradio ? @jpc @gongouveia

**Comment by BBC-Esq** [comment]
Just created this issue and am posting it here since it's kind of related.  Thanks.

[https://github.com/collabora/WhisperSpeech/issues/67](https://github.com/collabora/WhisperSpeech/issues/67)

**Comment by gongouveia** [comment]
@Josephrp Soz I have been quite buzy, I will try to add my layout contribution later today

**Comment by jpc** [comment]
Hey, I started to rework the demo a bit yesterday (including the UI). I should push a PR in a few hours.

**Comment by gongouveia** [comment]
Created a pull request in hf, to change layout.

**Comment by gongouveia** [comment]
@Josephrp demo is broken up, hope it was not my pr

**Comment by Josephrp** [comment]
I simply paused it to free up compute , as the official official version
was released here : https://huggingface.co/spaces/collabora/whisperspeech

I really learned a lot with you - so now… please check the whisperfusion
repo there’s a another issue for a demo there - I guess we can start 😉👊🏻

**Comment by Josephrp** [comment]
hey @gongouveia , i just started it again , it's building now, sorry about my confusion ! anyways i'll keep it up as long as i can :-) hope no harm done

**Comment by jpc** [comment]
Hey, sorry I wanted to put the demo online before my holiday and did not have a chance to review your layout changes.

I am back now but the updated Space is not working for me right now. @gongouveia have you seen the ["official" space](https://huggingface.co/spaces/collabora/WhisperSpeech)? If you have some ideas on how we could make it better I would love to merge your improvements.

**Comment by jpc** [comment]
I think that once we merge in MPS support we could get this app into the cocktail peanut's pinokio.computer

**Comment by gongouveia** [comment]
@jpc Wow, I had never heard of pinokio tool, very interesting..

Current version is looking good :)

**Comment by jpc** [comment]
Ok, thanks for all you help @Josephrp and @gongouveia! :)
